File: apps/hakka/Sources/Core/Rules/CapturedMockConverter.swift
import Foundation
import HakkaCommon

/// The capture → mock promotion: freezes a captured response into a mock
/// rule the device engines serve verbatim — replay the app's own real
/// response with no proxy in the path.
public enum CapturedMockConverter {
    /// Thrown by `entry(from:)` when the capture has nothing usable to
    /// replay — the request is still pending or the network call itself
    /// failed. Without this guard, promotion would fabricate a `200 ""`
    /// mock that silently masks the original failure.
    public enum PromotionError: Error, Equatable, LocalizedError {
        case incompleteCapture(url: String, underlying: String?)

        public var errorDescription: String? {
            switch self {
            case .incompleteCapture(let url, let underlying):
                if let underlying, !underlying.isEmpty {
                    return "Can't promote \(url) to a mock: the request failed (\(underlying))"
                }
                return "Can't promote \(url) to a mock: no response was captured"
            }
        }
    }

    /// A ready-to-install entry; re-promoting the same request replaces the
    /// same wire id rather than piling up duplicates. Throws
    /// `PromotionError.incompleteCapture` for a request with no status
    /// (pending, or dropped before a response arrived) or a recorded
    /// transport error — there is no real response to freeze.
    public static func entry(from request: NetworkRequest) throws -> RuleEntry {
        guard request.status != nil, request.error == nil else {
            throw PromotionError.incompleteCapture(url: request.url, underlying: request.error)
        }
        return RuleEntry(id: ruleID(for: request), payload: .mock(mockRule(from: request)))
    }

    /// Matches the endpoint, not one query string: scheme + host + port +
    /// path, with volatile query parameters dropped. Substring semantics —
    /// the engines match this against the full request URL.
    public static func mockRule(from request: NetworkRequest) -> MockRuleInput {
        let (headers, headerValues) = responseHeaders(from: request)
        return MockRuleInput(
            pattern: pattern(for: request),
            isRegex: false,
            method: request.method.rawValue,
            response: MockResponse(
                status: request.status ?? 200,
                headers: headers,
                headerValues: headerValues,
                body: request.responseBody ?? "",
                delay: 0
            ),
            enabled: true
        )
    }

    /// Deterministic id from the match key, so the second promotion of the
    /// same endpoint replaces the first. Wire-safe characters only.
    public static func ruleID(for request: NetworkRequest) -> String {
        let key = "\(request.method.rawValue) \(pattern(for: request))"
        var hash: UInt64 = 0xcbf2_9ce4_8422_2325
        for byte in key.utf8 {
            hash = (hash &- UInt64(byte)) &* 0x100_0000_01b3
        }
        return "mck-\(String(hash, radix: 36))"
    }

    private static func pattern(for request: NetworkRequest) -> String {
        guard let url = URL(string: request.url),
              let scheme = url.scheme?.isEmpty == false ? url.scheme : nil,
              let host = url.host else {
            return request.url
        }
        var pattern = "\(scheme)://\(host)"
        if let port = url.port {
            pattern += ":\(port)"
        }
        return pattern + url.path
    }

    /// The captured response headers that can be replayed verbatim. Bodies
    /// are stored decoded, so `Content-Encoding` would label plaintext as
    /// compressed, and `Content-Length` describes bytes the serving stack
    /// recomputes — both dropped; everything else (Content-Type, Set-Cookie,
    /// …) survives.
    ///
    /// Returns both halves of the wire's additive multi-value shape (see
    /// `MockResponse.headerValues`'s doc): `headers` carries one
    /// representative value per name (old decoders/consumers keep working
    /// unchanged), and `headerValues` carries the full ordered list for any
    /// name that had more than one captured value. `Set-Cookie` is the
    /// header this actually matters for — RFC 6265 §3 forbids combining
    /// multiple Set-Cookie values into one field, because a comma can
    /// legally appear inside a cookie's own `Expires` attribute, so folding
    /// would be ambiguous or outright corrupt. Ordinary headers with 2+
    /// values are still comma-joined into `headers` per RFC 7230 §3.2.2 (that
    /// fold is safe for them) and are NOT duplicated into `headerValues` —
    /// only genuinely multi-value replay (Set-Cookie today) needs the list.
    private static func responseHeaders(from request: NetworkRequest) -> (headers: [String: String], headerValues: [String: [String]]) {
        let excluded: Set<String> = ["content-encoding", "content-length", "transfer-encoding", "connection"]
        var headers: [String: String] = [:]
        var headerValues: [String: [String]] = [:]
        for (name, values) in request.responseHeaders {
            let lower = name.lowercased()
            guard !excluded.contains(lower), !values.isEmpty else { continue }
            if lower == "set-cookie" {
                headers[name] = values[0]
                if values.count > 1 { headerValues[name] = values }
            } else {
                headers[name] = values.joined(separator: ", ")
            }
        }
        return (headers, headerValues)
    }
}

File: packages/hakka-cli/src/mcp/tools/capturedMockConverter.ts
import type { MockRuleInput, NetworkRequest } from 'hakka-core'

/**
 * The capture → mock promotion: freezes a captured response into a mock
 * rule the device engines serve verbatim. Mirrors the Swift desktop app's
 * `apps/hakka/Sources/Core/Rules/CapturedMockConverter.swift` decisions
 * exactly (pattern shape, dropped headers, deterministic id) so the two
 * surfaces cannot disagree about what "promote this capture" means.
 */

const EXCLUDED_RESPONSE_HEADERS = new Set(['content-encoding', 'content-length', 'transfer-encoding', 'connection'])

const FNV_OFFSET = 0xcbf2_9ce4_8422_2325n
const FNV_PRIME = 0x100_0000_01b3n
const MASK64 = 0xffff_ffff_ffff_ffffn

/**
 * Matches the endpoint, not one query string: scheme + host + port + path,
 * with the query string dropped. Falls back to the raw url (query string
 * included) when it doesn't parse as an absolute URL — same fallback the
 * Swift converter uses.
 */
function patternFor(url: string): string {
  let parsed: URL
  try {
    parsed = new URL(url)
  } catch {
    return url
  }
  const scheme = parsed.protocol.replace(/:$/, '')
  if (!scheme || !parsed.hostname) return url

  let pattern = `${scheme}://${parsed.hostname}`
  if (parsed.port) pattern += `:${parsed.port}`
  return pattern + parsed.pathname
}

/**
 * Deterministic id derived from the match key (method + pattern), so
 * re-promoting the same endpoint replaces the existing rule instead of
 * piling up duplicates. Same FNV-derived wraparound hash as the Swift
 * converter's `ruleID(for:)`, so the two surfaces compute the same id for
 * the same capture. Wire-safe characters only (`mck-<base36>`).
 */
function ruleIdFor(method: string, pattern: string): string {
  const key = `${method} ${pattern}`
  let hash = FNV_OFFSET
  for (const byte of Buffer.from(key, 'utf8')) {
    hash = ((hash - BigInt(byte)) & MASK64) * FNV_PRIME
    hash &= MASK64
  }
  return `mck-${hash.toString(36)}`
}

/**
 * The captured response headers that can be replayed verbatim, split the
 * same way `MockResponse` is: one representative (comma-joined) value per
 * name in `headers`, plus the real ordered list in `headerValues` for any
 * name that arrived with more than one value. Bodies are stored decoded, so
 * `Content-Encoding` would mislabel plaintext as compressed and
 * `Content-Length`/`Transfer-Encoding` describe bytes the serving stack
 * recomputes — those (plus `Connection`) are dropped from both; everything
 * else (Content-Type, Set-Cookie, …) survives.
 *
 * Multi-value survival depends on the capture source populating
 * `NetworkRequest.responseHeaderValues` (see `model/types.ts`).
 * `hakka-node`'s `http`/`https` interceptor (`httpInterceptor.ts`) does this
 * for any name Node handed back as a real `string[]` — chiefly `Set-Cookie`,
 * where RFC 6265 §3 forbids folding multiple values into one comma-joined
 * field. A capture with no `responseHeaderValues` (e.g. `fetch`, whose
 * `Headers` API already collapses `Set-Cookie` per spec before user code
 * ever sees it) still promotes fine — it just carries the single folded
 * value in `headers`, same as before this widening existed. This now
 * matches `apps/hakka/Sources/Core/Rules/CapturedMockConverter.swift`'s
 * outcome for a Node-captured Set-Cookie response, even though the iOS
 * capture model (`[String: [String]]` from the start) and this one
 * (single-value `responseHeaders` plus an additive `responseHeaderValues`)
 * arrive at it by different routes.
 */
function responseHeadersFor(request: NetworkRequest): {
  headers: Record<string, string>
  headerValues?: Record<string, string[]>
} {
  const headers: Record<string, string> = {}
  for (const [name, value] of Object.entries(request.responseHeaders ?? {})) {
    if (EXCLUDED_RESPONSE_HEADERS.has(name.toLowerCase())) continue
    headers[name] = value
  }

  let headerValues: Record<string, string[]> | undefined
  for (const [name, values] of Object.entries(request.responseHeaderValues ?? {})) {
    if (EXCLUDED_RESPONSE_HEADERS.has(name.toLowerCase())) continue
    if (values.length < 2) continue
    headerValues ??= {}
    headerValues[name] = [...values]
  }

  return { headers, headerValues }
}

export type PromotionRefusalReason = 'errored_capture' | 'incomplete_capture'

/**
 * Refuses to promote a capture that never got a real response: an errored
 * capture (network error, timeout, abort) or one still pending (no status
 * yet). Promoting either would fabricate a nonsense `200 ""` mock instead
 * of reporting the truth — the defect this tool deliberately does not
 * reproduce.
 */
export function refusalReasonFor(request: NetworkRequest): PromotionRefusalReason | null {
  if (request.error) return 'errored_capture'
  if (request.status == null) return 'incomplete_capture'
  return null
}

/** Builds the mock rule + wire id for a captured request. Caller must check `refusalReasonFor` first. */
export function mockRuleEntryFor(request: NetworkRequest): MockRuleInput & { id: string } {
  const pattern = patternFor(request.url)
  const id = ruleIdFor(request.method, pattern)
  const { headers, headerValues } = responseHeadersFor(request)
  return {
    id,
    pattern,
    method: request.method,
    mode: 'mock',
    response: {
      status: request.status as number,
      headers,
      ...(headerValues !== undefined ? { headerValues } : {}),
      body: request.responseBody ?? '',
      delay: 0,
    },
    enabled: true,
  }
}
